pkg/controller/controller.go

```diff
@@ -1081,47 +1175,117 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+func (p *csiProvisioner) IsGranted(ctx context.Context, claim *v1.PersistentVolumeClaim) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		return false, nil
+	}
+
+	// Get all ReferenceGrants in snapshot's namespace
+	referenceGrants, err := p.referenceGrantLister.ReferenceGrants(*claim.Spec.DataSourceRef.Namespace).List(labels.Everything())
+	if err != nil {
+		return false, fmt.Errorf("error getting ReferenceGrants in %s namespace from api server: %v", *claim.Spec.DataSourceRef.Namespace, err)
+	}
+
+	var allowed bool
+	// Check that accessing to {namespace}/{name} is allowed.
+	for _, grant := range referenceGrants {
+		var validFrom bool
+		for _, from := range grant.Spec.From {
+			if from.Group == "" && from.Kind == pvcKind && string(from.Namespace) == claim.Namespace {
+				validFrom = true
+				break
+			}
+		}
+		// Skip unrelated policy by checking From field
+		if !validFrom {
+			continue
+		}
+
+		for _, to := range grant.Spec.To {
+			if claim.Spec.DataSourceRef.APIGroup != nil && (string(to.Group) != *claim.Spec.DataSourceRef.APIGroup || string(to.Kind) != claim.Spec.DataSourceRef.Kind) {
```

Comment:
What if `dataSourceRef.APIGroup == nil` and `to.Group is set`?

Also it should be `groupMatches` and `kindMatches` right?

Reply:
I fixed it as follows. Is this the fix you expected?
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R1145-R1149